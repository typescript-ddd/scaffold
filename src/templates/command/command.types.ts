import {
  TemplatePropertiesOption,
  TemplatePropertyOption,
  TemplateStringOption,
} from "../shared/template";

export type CommandTemplateValues = {
  entityName: string;
  actionName: string;
  properties: TemplatePropertyOption[];
};
export type CommandTemplateOptions = {
  entityName: TemplateStringOption;
  actionName: TemplateStringOption;
  properties: TemplatePropertiesOption;
};
