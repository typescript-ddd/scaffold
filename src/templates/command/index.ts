export * from "./command.generator";
export * from "./command.template";
export * from "./command.types";
