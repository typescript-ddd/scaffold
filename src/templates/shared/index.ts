export * from "./context-builder";
export * from "./generate-context";
export * from "./template";