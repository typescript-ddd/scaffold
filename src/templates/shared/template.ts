import { GenerateContext } from "./generate-context";

export type TemplateEnum = Record<string, string>;
export type TemplatePropertyValueType =
  | "string"
  | "number"
  | "boolean"
  | "date";

export type TemplateEnumOption = {
  name: string;
  label: string;
  description: string;
  enum: TemplateEnum;
  optionType: "enum";
};

export type TemplateStringOption = {
  name: string;
  label: string;
  description: string;
  optionType: "string";
};

export type TemplateTokenStringOption = {
  name: string;
  label: string;
  description: string;
  optionType: "tokenString";
};

export type TemplateBooleanOption = {
  name: string;
  label: string;
  description: string;
  optionType: "boolean";
};

export type TemplateNumberOption = {
  name: string;
  label: string;
  description: string;
  optionType: "number";
};

export type TemplateDateOption = {
  name: string;
  label: string;
  description: string;
  optionType: "date";
};

export type TemplatePropertyOption = {
  name: string;
  valueType: TemplatePropertyValueType;
  nullable?: boolean;
  prop?: string;
};

export type TemplatePropertyValue = {
  name: string;
  valueType: string;
  nullable?: boolean;
  prop?: string;
};

export type TemplatePropertiesOption = {
  name: string;
  label: string;
  description: string;
  optionType: "properties";
};

export type TemplateOption =
  | TemplateEnumOption
  | TemplateStringOption
  | TemplateBooleanOption
  | TemplateNumberOption
  | TemplateDateOption
  | TemplatePropertiesOption
  | TemplateTokenStringOption;

export type Maybe<T> = T | null | undefined;

export interface Template<
  TValues extends { [K in keyof TOptions]: any},
  TOptions extends { [key: string]: TemplateOption }
> {
  name: string;
  description: string;
  options: TOptions;
  defaultValues: TValues;
  generate(values: TValues, context: GenerateContext): string;
}