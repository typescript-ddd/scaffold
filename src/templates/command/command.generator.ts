import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { CommandTemplateValues } from "src/templates/command/command.types";

export const generateCommand = (
  values: CommandTemplateValues,
  context: GenerateContext
) => {
  const { entityName, actionName, properties } = values;
  const commandName = `${strings.capitalize(actionName)}${strings.capitalize(
    entityName
  )}Command`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${commandName} implements Command {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Command"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(commandName);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(commandName)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: [
      ...properties.map((property) => ({
        name: property.name,
        type: property.valueType,
        hasQuestionToken: property.nullable,
        scope: Scope.Public,
        isReadonly: true,
      })),
    ],
  });
  constructorDeclaration.setBodyText(`// TODO: Add validation`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      commandName
    )}.`,
    tags: properties.map((property) => ({
      tagName: "param",
      text: `{${property.valueType}} ${property.name} - The ${property.name}.`,
    })),
  });

  let output = sourceFile.getFullText();

  return output;
};
