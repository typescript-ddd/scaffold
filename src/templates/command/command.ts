import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateCommandOptions = {
  entityName: string;
  actionName: string;
  properties: {
    name: string;
    type: string;
    nullable?: boolean;
    description?: string;
  }[];
};

export const generateCommand = (
  options: GenerateCommandOptions,
  context: GenerateContext
) => {
  const { entityName, actionName, properties } = options;
  const commandName = `${strings.capitalize(actionName)}${strings.capitalize(
    entityName
  )}Command`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${commandName} implements Command {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Command"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(commandName);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(commandName)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: [
      ...properties.map((property) => ({
        name: property.name,
        type: property.type,
        hasQuestionToken: property.nullable,
        scope: Scope.Public,
        isReadonly: true,
      })),
    ],
  });
  constructorDeclaration.setBodyText(`// TODO: Add validation`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      commandName
    )}.`,
    tags: properties.map((property) => ({
      tagName: "param",
      text: `{${property.type}} ${property.name} ${
        property.description ? `- ${property.description}` : ""
      }`,
    })),
  });

  let output = sourceFile.getFullText();

  return output;
};
