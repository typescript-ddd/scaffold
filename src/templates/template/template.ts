import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateTemplateOptions = {
  name: string;
};

export const generateTemplate = (
  options: GenerateTemplateOptions,
  context: GenerateContext
) => {
  const { name } = options;
  const className = strings.capitalize(name);

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Entity"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `${strings.describe(name)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [{ name: "value", type: "string" }],
  });
  constructorDeclaration.setBodyText(`super(value);`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )} identifier.`,
    tags: [{ tagName: "param", text: "value The value of the identifier." }],
  });

  let output = sourceFile.getFullText();

  return output;
};
