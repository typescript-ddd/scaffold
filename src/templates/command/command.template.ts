import { Chunk, GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateCommand } from "./command.generator";
import { CommandTemplateOptions, CommandTemplateValues } from "./command.types";

export class CommandTemplate
  implements Template<CommandTemplateValues, CommandTemplateOptions>
{
  readonly name = "Command";
  readonly description = "Generates a command";
  readonly options: CommandTemplateOptions = {
    actionName: {
      name: "actionName",
      label: "Action Name",
      description: "The name of the action",
      optionType: "string",
    },
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
    properties: {
      name: "properties",
      label: "Properties",
      description: "The properties of the command",
      optionType: "properties",
    },
  };
  readonly defaultValues = {
    actionName: "Create",
    entityName: "User",
    properties: [],
  };
  generate(values: CommandTemplateValues, context: GenerateContext, chunkName?: string): Chunk {
    return generateCommand(values, context, chunkName);
  }
}
