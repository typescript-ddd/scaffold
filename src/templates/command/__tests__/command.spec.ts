import { ContextBuilder, GenerateContext } from "../../shared";
import { CommandTemplate } from "../command.template";

describe("Command", () => {
  let context: GenerateContext;
  let template: CommandTemplate;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
    template = new CommandTemplate();
  });

  it("Should render the command template", () => {
    const output = template.generate(
      {
        entityName: "User",
        actionName: "Create",
        properties: [
          { name: "name", valueType: "string", nullable: false },
          {
            name: "email",
            valueType: "string",
          },
        ],
      },
      context
    );

    expect(output).toMatchSnapshot();
  });
});
