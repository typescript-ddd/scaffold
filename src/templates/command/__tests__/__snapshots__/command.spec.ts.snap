// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command Should render the command template 1`] = `
"import { Command } from "@typescript-ddd/core";

/** Represents a create user command. */
export class CreateUserCommand implements Command {
    /**
     * Initializes a new instance of a create user command.
     * @param {string} name - The name.
     * @param {string} email - The email.
     */
    constructor(public readonly name: string, public readonly email: string) {
        // TODO: Add validation
    }
}    
"
`;
